llm-models/embedding/bge/bge-large/04_fine_tune_embedding.py.py

```diff
@@ -0,0 +1,188 @@
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ### Data preprocessing
+# MAGIC
+# MAGIC Convert the examples into InputExample's.
+
+# COMMAND ----------
+
+from sentence_transformers import InputExample
+
+def create_dataset_for_multiple_loss(train_dataset):
+  train_examples = []
+  for elem in train_dataset:
+    query = f"Represent this sentence for searching relevant passages: {elem['query']}"
+    texts = elem["pos"]
+    
+    for text in texts:
+      train_examples.append(InputExample(texts=[query, text]))
+  return train_example
+
+train_examples = create_dataset_for_multiple_loss(dataset)
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC
+# MAGIC ### Fine tune the embedding model
+# MAGIC
+# MAGIC Since we only have sentences pairs with no labels, we can use the `MultipleNegativesRankingLoss` function.
+
+# COMMAND ----------
+
+from sentence_transformers import SentenceTransformer
+model = SentenceTransformer('BAAI/bge-large-en')
+
+# COMMAND ----------
+
+from torch.utils.data import DataLoader
+from sentence_transformers import losses
+from accelerate import notebook_launcher
+
+train_dataloader = DataLoader(train_examples, shuffle=True, batch_size=3)
+train_loss = losses.MultipleNegativesRankingLoss(model=model)
+
+model.fit(train_objectives=[(train_dataloader, train_loss)], epochs=3)
+
+# COMMAND ----------
+
+model.save(output_model_path)
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ## Save fine tuned model to mlflow
+
+# COMMAND ----------
+
+# Save the model as Pyfunc to mlflow
+import mlflow
+import torch
+import pandas as pd
+from sentence_transformers import SentenceTransformer
+ 
+class SentenceTransformerEmbeddingModel(mlflow.pyfunc.PythonModel):
+  def load_context(self, context):
+    device = 0 if torch.cuda.is_available() else -1
+    self.model = SentenceTransformer(context.artifacts["sentence-transformer-model"], device=device)
+    
+  def predict(self, context, model_input): 
+    texts = model_input.iloc[:,0].to_list() # get the first column
+    sentence_embeddings = self.model.encode(texts)
+    return pd.Series(sentence_embeddings.tolist())
+
+# COMMAND ----------
+
+from mlflow.utils.environment import _mlflow_conda_env
+import accelerate
+import sentence_transformers
+import cloudpickle
+EMBEDDING_CONDA_ENV = _mlflow_conda_env(
+    additional_pip_deps=[
+        f"accelerate=={accelerate.__version__}",
+        f"cloudpickle=={cloudpickle.__version__}",
+        f"sentence-transformers=={sentence_transformers.__version__}",
+    ]
+)
+
+# COMMAND ----------
+
+import mlflow
+
+with mlflow.start_run() as run:
+  my_model = SentenceTransformerEmbeddingModel()
+  model_info = mlflow.pyfunc.log_model(artifact_path="model", python_model=my_model, input_example=["London is known for its finacial district"], artifacts={"sentence-transformer-model": output_model_path}, conda_env=EMBEDDING_CONDA_ENV)
```

Comment:
Do we want to add an additional query example with the prefix text "Represent this sentence for searching relevant passages: "?